compiler/src/iree/compiler/Codegen/Common/MaterializeTuningSpecsPass.cpp

```diff
@@ -197,14 +197,30 @@ struct MaterializeTuningSpecsPass final
       return;
     }
 
-    // If only the default tuning spec is available, use it directly and skip
-    // the linking stage.
-    if (!hasUserTuningSpec) {
-      if (failed(dumpFinalTuningSpecToDir(*defaultTuningSpec))) {
+    // Check if the user-provided tuning spec has the default entry point
+    // attribute.
+    bool userTuningSpecWithDefaultAttr =
+        hasUserTuningSpec &&
+        (*userTuningSpec)->hasAttr(kTuningSpecDefaultEntrypointAttrName);
+
+    // Determine if the linking pass should be skipped.
+    // Skip if there is a user-provided spec with the default attribute but no
+    // default tuning spec, or if there is no user-provided spec but a default
+    // tuning spec is available.
+    bool skipLinkPass = (hasUserTuningSpec && !hasDefaultTuningSpec &&
+                         userTuningSpecWithDefaultAttr) ||
+                        (!hasUserTuningSpec && hasDefaultTuningSpec);
```

Comment:
Instead of this logic, can we move up some code from down below to perform this check?
```c++
    SmallVector<ModuleOp, 2> allSpecs = {*userTuningSpec};
    if (hasDefaultTuningSpec) {
      allSpecs.push_back(*defaultTuningSpec);
    }
```
(and handle the missing user spec case too.)

Then we can check that there's a single element in this vector and that it has the default entrypoint attr.